### PR TITLE
Add menu divider, secondary line, and visual tweaks

### PR DESCRIPTION
**Pull request by FlyersPh9**
- Added menu divider.
- Added secondary line support.
- Visual changes to disabled line & focus.
- All sizes of select, text field, text area have matching horizontal padding so that the icon or label aligned with the menu.  This allows us to remove `$iui-component-padding-horizontal-small` and `$iui-component-padding-horizontal-large`.
- Special `.iui-header-menu-icon` modifier applied to `.iui-icon`, gives the extra spacing to align with header button's 24px thumbnail area.

Screen Shot 2021-06-30 at 5 10 45 PM

### CODE REVIEW

File: src/menu/menu.scss [hunk 1]
```diff
@@ -94,6 +119,15 @@
       }
     }
   }
+
+  > .iui-divider {
```
Comment (mayank99) [inline src/menu/menu.scss:123]: I would call this `iui-menu-divider` or something, and move it outside. It's specificity should be very low

File: src/menu/menu.scss [hunk 2]
```diff
@@ -94,6 +119,15 @@
...
+    height: 1px;
+    width: calc(100% - #{$iui-l});
```
Comment (mayank99) [inline src/menu/menu.scss:125]: You don't need `#{}`

File: src/menu/menu.scss [hunk 3]
```diff
@@ -58,13 +68,24 @@
...
+        outline: 1px solid t(iui-color-foreground-primary);
+        outline-offset: -1px;
+      }
+    }
+  
+    &:focus:not(:focus-visible) {
       outline: none;
     }
+  
+    &:focus-visible {
```
Comment (mayank99) [inline src/menu/menu.scss:81]: This is not necessary

File: src/menu/menu.scss [hunk 4]
```diff
@@ -78,13 +99,17 @@
           fill: t(iui-icons-color-primary);
         }
       }
+
+      &:focus,
+      &:focus-visible {
+        outline-offset: -2px;
```
Comment (mayank99) [inline src/menu/menu.scss:105]: use variable for 2px.

Also, `:focus, :focus-visible` breaks compatibility with older browsers. I don't think you even need focus-visible here

File: src/menu/menu.scss [hunk 5]
```diff
@@ -36,19 +37,28 @@
...
+        }
+      }
 
       + .iui-icon {
         margin-right: 0;
         margin-left: $iui-s;
       }
+
+      &,
+      > * {
```
Comment (mayank99) [inline src/menu/menu.scss:54]: Not sure about `*`

File: src/menu/menu.scss [hunk 6]
```diff
@@ -19,13 +19,14 @@
   margin: 0;
   padding: 0;
 
-  > li {
+  > li:not(.iui-divider) {
```
Comment (mayank99) [inline src/menu/menu.scss:22]: This increases the specificity 🤔

File: src/menu/menu.scss [hunk 7]
```diff
@@ -36,19 +37,28 @@
 
     > .iui-content {
       flex: 1 1 auto;
-      white-space: nowrap;
-      overflow: hidden;
-      text-overflow: ellipsis;
-      height: $iui-line-height;
+
+      .iui-menu-description {
```
Comment (mayank99) [inline src/menu/menu.scss:41]: I don't think this needs to be nested

File: src/menu/menu.scss [hunk 8]
```diff
@@ -19,13 +19,14 @@
   margin: 0;
   padding: 0;
 
-  > li {
+  > li:not(.iui-divider) {
     display: flex;
     align-items: center;
     font-size: $iui-font-size;
-    padding: 0 $iui-sm;
+    padding: $iui-component-padding-vertical $iui-sm;
```
Comment (mayank99) [inline src/menu/menu.scss:26]: How come it was 0 before?

**Review by mayank99 — COMMENTED** [review]
@FlyersPh9  Before any further review, can you verify the visuals? The spacing just looks off to me.

File: src/menu/menu.scss (hunk 2 above)
Comment (FlyersPh9) [reply src/menu/menu.scss:125]: I think it is required, notice the width of the dividing line.  With `#{ }`:
Screen Shot 2021-06-28 at 11 06 41 AM

Without `#{ }`:
Screen Shot 2021-06-28 at 11 07 08 AM

File: src/menu/menu.scss (hunk 1 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: Renamed to `iui-menu-divider`.  Also introduced `iui-menu-item` so I can target the class directly rather than targeting the `li` element.  This allows me to remove the part that says `li:not(.iui-menu-divider)`.

File: src/menu/menu.scss (hunk 3 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: Removed.

File: src/menu/menu.scss (hunk 4 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: Using the variable and removed focus-visible.

File: src/menu/menu.scss (hunk 6 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: Added a `.iui-menu-item` class so I can avoid using this `:not`.

File: src/menu/menu.scss (hunk 2 above)
Comment (mayank99) [reply src/menu/menu.scss:117]: Oh right, because it's inside `calc`

File: src/menu/menu.scss (hunk 1 above)
Comment (mayank99) [reply src/menu/menu.scss]: The specificity is still high. I'd move both of those things out into their own flat mixins, and declare in classes.scss

File: src/menu/menu.scss (hunk 6 above)
Comment (mayank99) [reply src/menu/menu.scss]: The specificity went from 0-1-1 to 0-2-1 to 0-2-0.

If you remove nesting, it will be 0-1-0.

File: src/menu/menu.scss (hunk 8 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: Reverted back to 0, using component height vars instead.  Does now require `.iui-large` to be applied to enable the secondary line within a menu item.

File: src/menu/menu.scss (hunk 7 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: Is no longer nested.

File: src/menu/menu.scss (hunk 5 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: Removed this, gave every nested div a class name instead.

File: src/menu/menu.scss (hunk 1 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: > @FlyersPh9 Before any further review, can you verify the visuals? The spacing just looks off to me.

Since replacing the padding with a fixed height variable, the menu item height should match the height of the medium & large select component.

**Comment by mayank99** [comment]
> > @FlyersPh9 Before any further review, can you verify the visuals? The spacing just looks off to me.
> 
> Since replacing the padding with a fixed height variable, the menu item height should match the height of the medium & large select component.

image

I think this looks even worse now 😄 There is more vertical space between the label and description than between them and the edges.

Also, what about the horizontal space between the icon and text? Probably should be higher for a larger menu item?

And what about that last menu item? I think that one looks out of place. Maybe the divider needs some margin above and below?

**Comment by FlyersPh9** [comment]
> I think this looks even worse now 😄 There is more vertical space between the label and description than between them and the edges. Looks better when there's no description though.

I am trying to match the line height used tabs. I inspected the CSS for tabs and didn't realize it was inheriting the browser default styling `line-height: normal`, so after applying that here it looks to match better.

Screen Shot 2021-06-28 at 3 07 18 PM

**Comment by mayank99** [comment]
Looks much better now (vertically).

I would suggest changing the right margin of icon from 8px to 12px (so it aligns with the left padding of menu item). I think that will make it look good on all sizes:
| 12px | 8px |
| --- | --- |
| image | image |

**Comment by FlyersPh9** [comment]
> I would suggest changing the right margin of icon from 8px to 12px (so it aligns with the left padding of menu item). I think that will make it look good on all sizes

I made some changes so the first icon is indented to match the indentation of the select component.  Screenshots:
Screen Shot 2021-06-29 at 9 51 30 AM
Screen Shot 2021-06-29 at 10 14 15 AM
Screen Shot 2021-06-29 at 10 14 04 AM

**Comment by bentleyvk** [comment]
> > I would suggest changing the right margin of icon from 8px to 12px (so it aligns with the left padding of menu item). I think that will make it look good on all sizes
> 
> I made some changes so the first icon is indented to match the indentation of the select component. Screenshots:
> Screen Shot 2021-06-29 at 9 51 30 AM
> Screen Shot 2021-06-29 at 10 14 15 AM
> Screen Shot 2021-06-29 at 10 14 04 AM

Is it really that bad to have the same spacing? For me it looks ok and requires less code. Only the dropdown button and the borderless select don't align with the menu but I don't think it would look nice with such small paddings.
image

**Comment by FlyersPh9** [comment]
> Is it really that bad to have the same spacing? For me it looks ok and requires less code. Only the dropdown button and the borderless select don't align with the menu but I don't think it would look nice with such small paddings.
> image

Latest push makes selects, text fields, text areas, of all sizes use the same horizontal padding so that icons & labels now align with the menu items.

File: src/header/header.scss [hunk 9]
```diff
@@ -219,6 +219,10 @@
 
     > .iui-icon {
       flex-shrink: 0;
+
+      &:not(.iui-header-button-icon) {
```
Comment (bentleyvk) [inline src/header/header.scss:223]: Are these changes really necessary for this PR?

File: src/inputs/input.scss [hunk 10]
```diff
@@ -21,12 +21,14 @@
   }
 
   &.iui-small {
-    padding: $iui-component-padding-vertical-small $iui-component-padding-horizontal-small;
+    padding-top: $iui-component-padding-vertical-small;
```
Comment (bentleyvk) [inline src/inputs/input.scss:24]: Why removed left and right padding?

File: src/menu/menu.scss [hunk 11]
```diff
@@ -19,89 +19,141 @@
...
+@mixin iui-menu-divider {
+  height: 1px;
+  width: calc(100% - #{$iui-l});
+  margin: auto;
+  @include themed {
+    background-color: t(iui-color-background-4);
+  }
+}
+
 @mixin iui-menu-icon {
```
Comment (bentleyvk) [inline src/menu/menu.scss:148]: Spacing is missing.
image

File: src/style/variables.scss [hunk 12]
```diff
@@ -13,8 +13,6 @@ $iui-component-height-small: $iui-baseline * 3 - $iui-component-offset * 2; // 2
 $iui-component-height-large: $iui-baseline * 5 - $iui-component-offset * 2; // 49px
 
 $iui-component-padding-horizontal: $iui-sm; // 12px
-$iui-component-padding-horizontal-small: $iui-sm * 0.5; // 6px
```
Comment (bentleyvk) [inline src/style/variables.scss:16]: Are these really not needed anymore?

**Comment by bentleyvk** [comment]
> > Is it really that bad to have the same spacing? For me it looks ok and requires less code. Only the dropdown button and the borderless select don't align with the menu but I don't think it would look nice with such small paddings.
> > image
> 
> Latest push makes selects, text fields, text areas, of all sizes use the same horizontal padding so that icons & labels now align with the menu items.

It is hard to tell whether these changes are good or not 😁 It is good because default size select, button and inputs content aligns vertically but other sizes don't so I am not sure if it is worth changing.
What do you think @veekeys @mayank99

**Comment by veekeys** [comment]
> > > Is it really that bad to have the same spacing? For me it looks ok and requires less code. Only the dropdown button and the borderless select don't align with the menu but I don't think it would look nice with such small paddings.
> > > image
> > 
> > 
> > Latest push makes selects, text fields, text areas, of all sizes use the same horizontal padding so that icons & labels now align with the menu items.
> 
> It is hard to tell whether these changes are good or not 😁 It is good because default size select, button and inputs content aligns vertically but other sizes don't so I am not sure if it is worth changing.
> Looked at other UI libs and they are not matching spacing of dropdown menu with button or other components only with select.
> What do you think @veekeys @mayank99

I think we should have added here only divider and secondary line.. not change the sizes of it and other components too...
If that is really needed, we could do it in separate PR. Now if its really needed, I dont know. Would need to research and hear @FlyersPh9 cent here..

File: src/style/variables.scss (hunk 12 above)
Comment (FlyersPh9) [reply src/style/variables.scss:16]: They are needed if we want different left / right padding for small, medium, large sizes.  If we choose to have all 3 sizes match like the PR does currently, the large & small variables aren't needed anymore.

File: src/menu/menu.scss (hunk 11 above)
Comment (FlyersPh9) [reply src/menu/menu.scss:148]: Yikes, that's embarrassing.  It is fixed with the next push.

File: src/header/header.scss (hunk 9 above)
Comment (FlyersPh9) [reply src/header/header.scss:223]: We could remove this if it will help it pass.  Without this, the icons do not align.  We also won't be able to show a photo thumbnail, we'll only have icons for now.

**With this CSS:**
header-menu-with

**Without this CSS:**
header-menu-without

Worth pointing out tho that [the initial request / bug filed](https://github.com/iTwin/iTwinUI/issues/85) is using the double line menu items from the context of the header button.

File: src/inputs/select.scss [hunk 13]
```diff
@@ -28,7 +28,7 @@ $iui-after-border: 5px solid;
       @include iui-menu-icon;
     }
 
-    > .iui-content {
+    > .iui-label {
```
Comment (mayank99) [inline src/inputs/select.scss:31]: Why was this renamed?

File: src/menu/menu.scss [hunk 14]
```diff
@@ -41,6 +39,7 @@
   cursor: pointer;
   box-sizing: border-box;
   line-height: normal;
+  border: 1px solid transparent;
```
Comment (mayank99) [inline src/menu/menu.scss:42]: Flashbacks from #80 😬

File: backstop/tests/menu.html [hunk 15]
```diff
@@ -29,162 +40,1879 @@
...
+            class="iui-icon"
+            viewBox="0 0 16 16"
+            aria-hidden="true"
+          >
+            <path
+              d="m6.00263 4.80681v6.39946a.27938.27938 0 0 0 .44292.24059l3.45479-3.17567a.34022.34022 0 0 0 .00046-.4807l-.00046-.00044-3.45479-3.22382a.269.269 0 0 0 -.44292.24058z"
+            />
+          </svg>
+        </li>
+        <li class="iui-menu-divider"></li>
```
Comment (mayank99) [inline backstop/tests/menu.html:332]: Add `aria-hidden` to dividers

File: backstop/tests/menu.html [hunk 16]
```diff
@@ -29,162 +40,1879 @@
...
-              dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex
-              ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu
-              fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt
-              mollit anim id est laborum.
-            </span>
-            <!-- Use checkmark.svg from @itwin/itwinui-icons package. -->
+            <span class="iui-label">App name</span>
+          </div>
+
+          <div class="iui-divider"></div>
```
Comment (mayank99) [inline backstop/tests/menu.html:551]: Oh, I forgot about this class 😄

File: src/inputs/select.scss [hunk 17]
```diff
@@ -68,7 +68,7 @@ $iui-after-border: 5px solid;
       }
     }
 
-    &.iui-active {
+    &.iui-expanded {
```
Comment (mayank99) [inline src/inputs/select.scss:71]: Do not like this

File: src/menu/menu.scss [hunk 18]
```diff
@@ -19,89 +19,140 @@
...
+
+  .iui-label,
+  .iui-menu-label,
+  .iui-menu-description {
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
+  }
+
+  .iui-label {
```
Comment (mayank99) [inline src/menu/menu.scss:55]: Need to avoid these generic names

**Review by mayank99 — COMMENTED** [review]
Looking a lot better now! But I don't like some of the class names

File: src/inputs/select.scss (hunk 13 above)
Comment (FlyersPh9) [reply src/inputs/select.scss:31]: Button, select, menu, header button are now all aligned to use `.iui-icon + .iui-label`.  We could re-use the `@include iui-menu-icon` across more components now.

File: backstop/tests/menu.html (hunk 15 above)
Comment (FlyersPh9) [reply backstop/tests/menu.html]: `aria-hidden="true"` added to all `.iui-menu-divider`s.

File: src/inputs/select.scss (hunk 13 above)
Comment (bentleyvk) [reply src/inputs/select.scss:31]: I also don't like this rename. Because now we have `label` inside `label`, `content` makes more sense here.
image

File: backstop/scenarios/menu.js [hunk 19]
```diff
@@ -1,7 +1,26 @@
-const { scenario } = require('../scenarioHelper');
+const { scenario, hover, focus } = require('../scenarioHelper');
 
 module.exports = [
-  scenario('basic', {
+  // Menu types
+  scenario('Type Default', {
+    selectors: ['#demo-default'],
+  }),
+  scenario('Type Header', {
```
Comment (bentleyvk) [inline backstop/scenarios/menu.js:8]: Do we really want to have tests for context? Because I really like this idea of having context but maybe don't flood our tests.

File: src/style/variables.scss (hunk 12 above)
Comment (bentleyvk) [reply src/style/variables.scss:16]: I asked because you have used multiplication in buttons:
image

So maybe we need to have `small` and `large` variables here just to avoid different sizing in the future.

File: src/header/header.scss (hunk 9 above)
Comment (bentleyvk) [reply src/header/header.scss:223]: Huh, how this affects alignment?

File: src/inputs/select.scss (hunk 17 above)
Comment (bentleyvk) [reply src/inputs/select.scss:71]: Don't start refactoring...

File: backstop/tests/menu.html [hunk 20]
```diff
@@ -29,162 +40,1915 @@
...
+          </span>
+        </li>
+      </ul>
+
+    </section>
+
+    <hr />
+
+    <h2>In context</h2>
+    <h3>Header</h3>
```
Comment (bentleyvk) [inline backstop/tests/menu.html:351]: Does not align.
image

File: src/button/default.scss [hunk 21]
```diff
@@ -43,6 +43,15 @@
     &.iui-large {
       padding-right: $iui-sm;
     }
+
+    // Flipping the expand arrow
```
Comment (bentleyvk) [inline src/button/default.scss:47]: Unrelated with current PR.

File: src/inputs/select.scss (hunk 13 above)
Comment (FlyersPh9) [reply src/inputs/select.scss]: Reverted back to `.iui-content` for the select component.  It should probably be renamed in another PR to more closely align with `.iui-menu-item-content`.

The menu icon mixin is now targeting `[class^='iui-'][class$='-content']` which should effect `.iui-menu-item-content`, `.iui-content` (as it is now), and `.iui-select-content` if / when we decide to rename that class.

File: src/inputs/select.scss (hunk 17 above)
Comment (FlyersPh9) [reply src/inputs/select.scss]: Reverted class name change and removed addition of `.iui-expanded` from button and header button.

File: src/menu/menu.scss (hunk 18 above)
Comment (FlyersPh9) [reply src/menu/menu.scss]: Renamed to `.iui-menu-item-content`.

File: src/button/default.scss (hunk 21 above)
Comment (FlyersPh9) [reply src/button/default.scss]: Removed all of this.

File: backstop/scenarios/menu.js (hunk 19 above)
Comment (FlyersPh9) [reply backstop/scenarios/menu.js:8]: I've removed the tests for selects & dropdown buttons because the menu itself was the same as the default test.  I left the header test in because it is styled slightly different to get the image / icon to align properly.

File: backstop/tests/menu.html (hunk 20 above)
Comment (FlyersPh9) [reply backstop/tests/menu.html:351]: I can't figure out why the test images have the menu's not aligned properly with the header buttons.  They look fine in the demo before:
Screen Shot 2021-07-07 at 4 37 26 PM

They menus are placed using `margin-left` inline and then some random value to make it look like they align.

File: src/header/header.scss (hunk 9 above)
Comment (FlyersPh9) [reply src/header/header.scss:223]: Sorry, I must have misunderstood what the initial comment was questioning.

I've removed this CSS since we also removed `.iui-expanded` from the header button.

File: src/style/variables.scss (hunk 12 above)
Comment (FlyersPh9) [reply src/style/variables.scss:16]: I've added those variables back in exclusively for the button to use now.

File: backstop/tests/menu.html (hunk 20 above)
Comment (bentleyvk) [reply backstop/tests/menu.html:351]: For me it is not only in tests but also when I open HTML

Comment (bentleyvk) [reply backstop/tests/menu.html:352]: Fixed menu placement, for some reason margin should change based on screen size.

File: src/inputs/select.scss (hunk 13 above)
Comment (bentleyvk) [reply src/inputs/select.scss]: I reverted back to `iui-content`. I hope Jon won't be mad about this 🤞

**Review by bentleyvk — APPROVED** [review]

File: src/inputs/select.scss (hunk 13 above)
Comment (FlyersPh9) [reply src/inputs/select.scss]: I'm not mad, I'm just happy this PR is finally about to pass! ☺️

**Comment by mayank99** [comment]
> I would suggest changing the right margin of icon from 8px to 12px (so it aligns with the left padding of menu item). I think that will make it look good on all sizes:
> 
> | 12px | 8px |
> | --- | --- |
> | image | image |

So we don't want to make this change? And yet we are changing all the buttons for no reason?




Also we need to fix multi-select borders at some point.
image

**Review by mayank99 — COMMENTED** [review]
I added some JS to toggle selection (it's not perfect but it does the job) and found a bug with focus. Other than that, this PR is good to go!

File: src/menu/menu.scss [hunk 22]
```diff
@@ -18,90 +18,148 @@
...
     }
+  }
 
-    > .iui-icon {
-      @include iui-menu-icon;
+  > .iui-icon {
+    @include iui-menu-icon;
+  }
+
+  &:hover:not(.iui-disabled):not(.iui-active) {
```
[the remark this answers is not in the document]
Comment (FlyersPh9) [reply src/menu/menu.scss:80]: I moved `iui-disabled` above this line.  I am able to simplify the selector to `&:hover:not(.iui-disabled)`.  If I remove the not disabled part, you'll see the background color change on hover.
Screen Shot 2021-07-12 at 9 48 05 AM

[the remark this answers is not in the document]
Comment (FlyersPh9) [reply src/menu/menu.scss:80]: Oh no, wait… if I remove `:not(.iui-active)` the hover overrides the active state.  
hover

[the remark this answers is not in the document]
Comment (mayank99) [reply src/menu/menu.scss:80]: 😦 Okay, I guess leave it

File: src/button/button.scss [hunk 23]
```diff
@@ -62,23 +62,11 @@ $iui-button-padding-large: $iui-xs * 6;
...
 
   &.iui-large {
     padding: 0 $iui-button-padding-large;
     height: $iui-component-height-large;
     font-size: $iui-font-size-leading;
-
-    > .iui-icon + .iui-label,
-    > .iui-label + .iui-icon,
-    > .iui-icon + .iui-icon {
-      margin-left: $iui-sm;
```
[the remark this answers is not in the document]
Comment (FlyersPh9) [reply src/button/button.scss]: Latest push reverts this.

**Review by mayank99 — APPROVED** [review]
